src/vm/runners/cairo_runner.rs

```diff
@@ -170,6 +170,25 @@ impl CairoRunner {
             }
         }
 
+        let inserted_builtins = builtin_runners
+            .iter()
+            .map(|x| &x.0)
+            .collect::<HashSet<&String>>();
+        let program_builtins: HashSet<&String> =
+            self.program.builtins.iter().collect::<HashSet<&String>>();
+        // Get the builtins that belong to the program but weren't inserted (those who dont belong to the instance)
+        let difference: HashSet<&&String> =
+            program_builtins.difference(&inserted_builtins).collect();
+        if !difference.is_empty() {
+            return Err(RunnerError::NoBuiltinForInstance(
+                difference
+                    .iter()
+                    .map(|x| (&(**x).clone()).clone())
+                    .collect(),
+                self.layout._name.clone(),
+            ));
+        }
+
```

Comment:
The following is more memory efficient as it avoids creation of a third `HashSet`:
```suggestion
        let difference = program_builtins
            .difference(&inserted_builtins)
            .peekable();
        if difference.peek().is_some() {
            return Err(RunnerError::NoBuiltinForInstance(
                difference
                    .map(|x| (&(**x).clone()).clone())
                    .collect(),
                self.layout._name.clone(),
            ));
        }

```

Actually we can do even better:
```suggestion
        if !inserted_builtins.is_subset(&program_builtins) {
            return Err(RunnerError::NoBuiltinForInstance(
                program_builtins
                    .difference(&inserted_builtins)
                    .map(|x| (&(**x).clone()).clone())
                    .collect(),
                self.layout._name.clone(),
            ));
        }

```

Reply:
Added suggested change(971773), but swapped` inserted_builtins.is_subset(&program_builtins)` for `program_builtins.is_subset(&inserted_builtins)`, as we want to make sure all the builtins in the compiled program are part of (or subset of) the inserted builtins

Reply:
Right, I misplaced those.